WebGPU integration

reeselevine opened the pull request:
Adds initial WebGPU integration to wllama. There are still a few issues to be resolved, but opening this PR now so we can start working on them!

To add WebGPU support, this PR does the following:

* Adds the WebGPU backend to the single thread WASM build, and adds a configuration option to wllama. If WebGPU is requested and available, it uses the WebGPU backend in the single thread WASM build. Otherwise, it uses either the multi-thread build or single thread build with the CPU backend.
* Instead of having consumers of wllama set `n_gpu_layers`, only the backend is specified in the config, and if WebGPU is requested/available, `n_gpu_layers` is set to 999. (This is how I've been running GPU backends locally, is this the right way to do it?)
* A few build flags and some code had to be changed to account for the WebGPU asynchronous operations. In particular, some wllama functions had to be included in `JSPI_EXPORTS` so that they are properly handled. `{'async' : true}` also has to be passed to `Module.cwrap` for these functions. We also found that it was necessary to export `HEAPU8`, even for CPU-only builds (maybe this is due to updating emscripten versions?).
* I also added a Github action to deploy to Github Pages, so that anyone can easily host `examples/main`. For example, I have it here: https://reeselevine.github.io/wllama/

## Issues to be resolved
* Wllama still does not work with WASM 64-bit builds, which is a problem because ggml's backend [get_memory](https://github.com/ggml-org/llama.cpp/blob/master/ggml/src/ggml-backend-impl.h#L148) function uses a `size_t` for free/total, which in WASM 32-bit builds overflows when the available memory is > 2^32 - 1 bytes. This is a problem on many machines, e.g., my M3 reports 2^32 bytes. So, in the built wasm in this PR, I actually hardcoded the return values from `get_memory` to 2^32 - 1. To address this issue, either (1) we need wllama to support 64-bit builds, which is not trivial (some interfaces between JavaScript/WASM need to change I think) and might come with other issues, or (2), change the ggml interface here to not use a size_t, or report the available memory in something other than bytes. This also gets to another question, which is that I haven't yet looked into how wllama is storing the models in memory, and if this can be optimized to avoid storing the model on the CPU side in memory if it's running on the GPU, or if it can currently support models larger than 4 GB on machines with more memory.
* Not all models work with the WebGPU backend yet. For example, some in the model-picker list fail because of overlapping buffer ranges in WebGPU. This is an issue I'm aware of, I just haven't had time yet to implement fixes in llama.cpp, since it requires special compilation of shaders with overlapping buffer ranges.
* There are some UI improvements in the example I'd like to add, as well as reporting of prefill/decode speeds (like WebLLM does).
* The code is not yet as performant as WebLLM. I think we can get there, and I think llama.cpp has advantages over WebLLM in being able to scalably provide performance across a bunch of different machines, and this is a long term project that I'm working on. On the bright side, on my machine it is much faster than using the CPU-only, using models like https://huggingface.co/unsloth/Llama-3.2-1B-Instruct-GGUF.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * WebGPU GPU acceleration for improved model inference performance with automatic CPU fallback
  * Real-time performance metrics displaying tokens/second and detailed timing information
  * Performance context visibility including token counts and cache reuse data
  * WebGPU hardware acceleration status indicator in the model interface
  * Performance context reset capability for benchmarking and testing

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

boingboomtschak committed:
Changes to get llama.cpp WebGPU backend working in build process

boingboomtschak committed:
Removing __pycache__

boingboomtschak committed:
Bumping emsdk image ver to 4.0.20 from 4.0.10, for newer included emdawnwebgpu port, removing remote port file from repo

boingboomtschak committed:
Adding WebGPU backend build to both single and multi threaded paths

boingboomtschak committed:
Explicitly exporting HEAPU8 (required by newer Emscripten ver)

boingboomtschak committed:
Pushing 64-bit flag change, submodule update, some JSPI/Asyncify investigation

reeselevine committed:
Separate path for webgpu build

reeselevine committed:
working

reeselevine committed:
Move to passing backend choice to wasm

reeselevine committed:
Cleanup

reeselevine committed:
Update flags

reeselevine committed:
deploy

reeselevine committed:
Add webgpu types to main example

reeselevine committed:
Add performance tracking

reeselevine committed:
Add webgpu support query to wllama api, address some issues

reeselevine commented on examples/main/src/utils/wllama.context.tsx:
```diff
@@ -6,7 +6,11 @@ import {
 import { Model, ModelManager, Wllama } from '@wllama/wllama';
-import { DEFAULT_INFERENCE_PARAMS, WLLAMA_CONFIG_PATHS } from '../config';
+import {
+  DEFAULT_BACKEND,
+  DEFAULT_INFERENCE_PARAMS,
+  WLLAMA_CONFIG_PATHS,
+} from '../config';
```
updated to use a boolean flag instead for preferring webgpu

reeselevine commented on src/workers-code/llama-cpp.js:
```diff
@@ -268,14 +269,14 @@ onmessage = async (e) => {
-        wllamaDebug = callWrapper('wllama_debug', 'string', []);
+        ], true);
+        wllamaExit = callWrapper('wllama_exit', 'string', [], false);
+        wllamaDebug = callWrapper('wllama_debug', 'string', [], false);
```
not sure this is true, at least for wllamaMalloc, despite it being async due to callWrapper returning an async function, the underlying code does not need async, i.e., passed to `Module.cwrap`, and in fact an error about `ret.then()` not being defined occurs if wllamaMalloc is called with `async: true`. I don't think `wllamaExit` or `wllamaDebug` are used in `examples/main` either though, so haven't fully tested this.

reeselevine committed:
Merge remote-tracking branch 'origin/perf'

reeselevine committed:
Update wasm blob with max(4gb, supported size)

reeselevine commented:
^ latest commits added a new API to wllama that surfaces performance data (tokens/second), and for now, I updated the wasm to support a max size of 4GB, or whatever the machine supports

ngxson commented:
Sorry for the late response, I'll try to look into it this week.

Re. 32/64-bit builds, I think we can try to fix it in a dedicated PR. Some parts of the code currently assume the address to be 32-bit (noticeably the code to "stream" the file to wasm memory context). But it shouldn't be too complicated to change then to 64-bit.

My main concern is that 64-bit can potentially break safari, but we can probably consider dropping safari support, or force using single-thread build if needed. Also, I think single-thread build won't be benefited from 64-bit (larger models), so we can probably constraint 64-bit to only be available on multi-threaded build

reeselevine commented:
No worries. As for single-thread vs. multi-thread 64-bit, when you review you'll see that I only added WebGPU support to the single-thread build, because the increased performance comes from offloading to the GPU, not from running multiple threads on the CPU. For example, on my machine I get ~32 t/s decode right now with the single-thread WebGPU backend, and ~10 t/s decode with the multi-thread CPU backend on a `Llama-3.2-1B-Instruct-Q4_0` model.

So we might want to come up with a plan for what wasm backends wllama wants to support, without having too many of them. I'm not sure what the full matrix of support in browsers is for WebGPU/multi-thread, but in my mind it definitely makes sense to have a single-thread 32-bit build fallback, with or without WebGPU support. Then the question is whether a multi-thread 64-bit build is necessary, or if a single-thread 64-bit build with WebGPU support should be enough for most usecases.

ngxson commented:
What I'm wondering whether multi-thread + webgpu improve performance. I suspect ops that are not supported by webgpu will be benefit from that, as they will be fallback to GPU implementation.

But I agree that this can be a later improvement. For now, I think having just 3 builds can be enough:
- single-thread CPU
- multi-thread CPU
- webgpu (currently single-thread, but can be multi in the future). we can also enable 64-bit only on this build

Edit: see my comment below

ngxson commented on .github/workflows/deploy-examples-main.yml:
```diff
@@ -0,0 +1,36 @@
+        working-directory: examples/main
+
+      - name: Deploy examples/main to gh-pages
+        uses: JamesIves/github-pages-deploy-action@v4
```
In case of using a 3rd-party action, it's recommended to pin the exact SHA commit: https://docs.github.com/en/actions/reference/security/secure-use#using-third-party-actions

reeselevine replied:
updated to use a specific commit

ngxson commented on cpp/glue.hpp:
```diff
@@ -494,7 +494,8 @@ struct glue_msg_load_req
   GLUE_FIELD(bool, n_ctx_auto)
   GLUE_FIELD(bool, use_mmap)
   GLUE_FIELD(bool, use_mlock)
-  GLUE_FIELD(int, n_gpu_layers)
```
I think `n_gpu_layers` still need to be kept, and reserved for future usage (for example, control how many layers that will be offloaded to GPU)

For now, for the webgpu case, we can hard-code it to 999 on JS code, whenever we set `use_webgpu = true`

reeselevine replied:
sounds good, added it back in

ngxson commented on scripts/docker-compose.yml:
```diff
@@ -19,23 +19,21 @@ services:
-
-        emcmake cmake ../..
-        export EMCC_CFLAGS="$$SHARED_EMCC_CFLAGS"
+        emcmake cmake ../.. -DGGML_WEBGPU=ON -DGGML_WEBGPU_JSPI=1 -DLLAMA_WASM_MEM64=0
```
just wondering if we can set these options inside [CMakeLists.txt](https://github.com/ngxson/wllama/blob/master/CMakeLists.txt) instead.

my long-term plan is to move all options there

reeselevine replied:
updated to set these options there

ngxson reviewed (commented):
looking again at the change, it seems there's no harm adding webgpu support to all the builds. so I think the only consideration here is whether to enable 64-bit

I think we can distribute 2 x 32-bit builds (single/multi-thread) plus one extra 64-bit build (can be single or multi depending on which one works better on webgpu)

reeselevine committed:
Add back n_gpu_layers, update builds

reeselevine committed:
formatting

reeselevine commented:
I added WebGPU to the multi-thread build. However, I found that using the multi-thread build led to a pretty significant performance decrease running in my browser. I don't fully understand how llama.cpp routes things in multi-thread mode yet, but I know the WebGPU backend has some inefficiencies/issues with that right now. It's something we definitely need to work on/fix going forwards in llama.cpp.

I imagine this can be improved, and like you said would allow non-GPU operations to run more efficiently, but for now I disabled using the multi-thread build with WebGPU and added a TODO in the code to look into this: https://github.com/ngxson/wllama/pull/198/files#diff-dc8126d70d08a19299c727046d114b88791f7fecac5c7735b905ec184bb07436R619. 

Otherwise, to support 32-bit builds and memory limits, I just opened this PR in llama.cpp: https://github.com/ggml-org/llama.cpp/pull/18707. The wasm blobs here are built against that code, once that solution or something else we come up with is merged in llama.cpp I'll update the submodule here.

ngxson reviewed (approved):
Looking good! We can merge when https://github.com/ggml-org/llama.cpp/pull/18707 is upstream and blobs are rebuilt.

One thing to note is that I will also need to design an API for lazy-loading tensors. Currently, tensors already offloaded to GPU will still have a copy inside wasm memory, which is wasteful.

reeselevine committed:
Update llama.cpp submodule and rebuild wasm blobs

SuperPauly commented:
> Looking good! We can merge when [ggml-org/llama.cpp#18707](https://github.com/ggml-org/llama.cpp/pull/18707) is upstream and blobs are rebuilt.
> 
> One thing to note is that I will also need to design an API for lazy-loading tensors. Currently, tensors already offloaded to GPU will still have a copy inside wasm memory, which is wasteful.

Looks like llama.cpp #18707 has been merged to main. Can this PR be merged now?

ngxson commented:
Not quite sure why CI is not run on PRs, I didn't notice that until now. Unfortunately the CI is current fails due to this error:

<img width="553" height="120" alt="image" src="https://github.com/user-attachments/assets/e7889e10-f0dc-433e-9ea3-13f784ff745e" />

And `npm run build` fails:

<img width="669" height="226" alt="image" src="https://github.com/user-attachments/assets/886470df-9803-4496-8ea4-7de83e794c93" />

I'll see if I can fix this quickly

ngxson commented:
@reeselevine I cannot push to your branch because it was created from your `master`, can you create a new PR from another branch?

reeselevine commented:
New PR opened here: https://github.com/ngxson/wllama/pull/201
